File: solutions/elixir/perfect-numbers/1/lib/perfect_numbers.ex
defmodule PerfectNumbers do
  @doc """
  Determine the aliquot sum of the given `number`, by summing all the factors
  of `number`, aside from `number` itself.

  Based on this sum, classify the number as:

  :perfect if the aliquot sum is equal to `number`
  :abundant if the aliquot sum is greater than `number`
  :deficient if the aliquot sum is less than `number`
  """
  @spec classify(number :: integer) :: {:ok, atom} | {:error, String.t()}
  def classify(number)
      when number < 1,
      do: {:error, "Classification is only possible for natural numbers."}

  def classify(number)
      when number == 1,
      do: {:ok, :deficient}

  def classify(number),
    do: find_number_type(number)

  defp find_number_type(number),
    do:
      number
      |> find_factors()
      |> Enum.sum()
      |> find_number_type(number)

  defp find_number_type(sum, number)
       when sum < number,
       do: {:ok, :deficient}

  defp find_number_type(sum, number)
       when sum == number,
       do: {:ok, :perfect}

  defp find_number_type(sum, number)
       when sum > number,
       do: {:ok, :abundant}

  defp find_factors(number),
    do:
      1..(number - 1)
      |> Enum.filter(&is_factor(number, &1))

  defp is_factor(target, candidate),
    do: rem(target, candidate) === 0
end

File: solutions/scala/perfect-numbers/1/src/main/scala/PerfectNumbers.scala
import NumberType.NumberType

object NumberType extends Enumeration {
  type NumberType = Value
  val Perfect, Abundant, Deficient = Value
}

object PerfectNumbers {
  def classify(integer: Int): Either[String, NumberType] =
    integer match {
      case value if value < 1 =>
        Left("Classification is only possible for natural numbers.")

      case value if value == 1 =>
        Right(NumberType.Deficient)

      case value if value > 1 =>
        Right(findNumberType(integer))
    }

  private def findNumberType(integer: Int) =
    this.findFactors(integer).sum match {
      case sum if sum < integer =>
        NumberType.Deficient

      case sum if sum == integer =>
        NumberType.Perfect

      case sum if sum > integer =>
        NumberType.Abundant
    }

  private def findFactors(integer: Int): List[Int] =
    (1 until (integer - 1))
      .toList
      .filter(this.isFactor(integer, _))

  private def isFactor(target: Int, candidate: Int): Boolean =
    target % candidate == 0
}

